File: homeworks/homework2.2/src/ImageProcessor.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class ImageProcessor {
    private static final int NUM_ROWS_OR_NUM_PIXELS_PER_ROW = 16;
    private final String inputFilename;
    private final String dir;
    private final List<List<Color>> image;

    public ImageProcessor(String inputFilename, String dir) {
        this.inputFilename = inputFilename;
        this.dir = dir;
        this.image = new ArrayList<>();
        readImage();
    }

    private void readImage() {
        try (Scanner scanner = new Scanner(new File(dir + inputFilename))) {
            int r_count = 0;
            while (scanner.hasNextLine()) {
                String line = scanner.nextLine();
                String[] pixels = line.split(" ");
                if (pixels.length != NUM_ROWS_OR_NUM_PIXELS_PER_ROW) {
                    throw new RuntimeException("Invalid amount of pixels in line " + r_count);
                }

                List<Color> row = getColors(pixels, r_count);
                image.add(row);
                r_count++;
            }
            if (r_count != NUM_ROWS_OR_NUM_PIXELS_PER_ROW) {
                throw new RuntimeException("Invalid amount of lines in file");
            }
        } catch (FileNotFoundException e) {
            throw new RuntimeException("Error opening file " + inputFilename);
        }
    }

    private static List<Color> getColors(String[] pixels, int r_count) {
        List<Color> row = new ArrayList<>();
        for (String pixel : pixels) {
            String[] rgb = pixel.split(",");
            if (rgb.length != 3) {
                throw new RuntimeException("Invalid pixel in line " + r_count);
            }
            try {
                Color color = new Color(Integer.parseInt(rgb[0]), Integer.parseInt(rgb[1]), Integer.parseInt(rgb[2]));
                if (!color.isValid()) {
                    throw new RuntimeException("Invalid pixel in line " + r_count);
                }
                row.add(color);
            } catch (NumberFormatException e) {
                throw new RuntimeException("Invalid pixel in line " + r_count);
            }
        }
        return row;
    }

    public void processImage(Color favoriteColor, Color unfavoriteColor) {
        for (int row = 0; row < image.size(); row++) {
            for (int col = 0; col < image.get(row).size(); col++) {
                Color currentColor = image.get(row).get(col);
                if (currentColor.equals(favoriteColor)) {
                    if (row > 0) image.get(row - 1).set(col, favoriteColor);
                    if (col > 0) image.get(row).set(col - 1, favoriteColor);
                } else if (unfavoriteColor != null && currentColor.equals(unfavoriteColor)) {
                    image.get(row).set(col, favoriteColor);
                }
            }
        }
    }

    public void writeImage(String outputFilename) throws IOException {
        BufferedWriter bw = new BufferedWriter(new FileWriter(dir + outputFilename));
        
        for (List<Color> row : image) {
            List<String> colorStrings = new ArrayList<>();
            for (Color color : row) {
                colorStrings.add(color.r + "," + color.g + "," + color.b);
            }
            bw.write(String.join(" ", colorStrings));
            bw.newLine();
        }
        
        bw.close();
    }
}

File: practices/practice2.2/src/ImageProcessor.cpp
#include "ImageProcessor.hpp"
#include <fstream>
#include <sstream>
#include <ranges>

bool Color::isValid() const {
    return r >= 0 && r <= 255 && g >= 0 && g <= 255 && b >= 0 && b <= 255;
}

bool operator==(const Color& a, const Color& b) {
    return a.r == b.r && a.g == b.g && a.b == b.b;
}

ImageProcessor::ImageProcessor(const std::string& inputFilename) : inputFilename(inputFilename) {
    readImage();
}

void ImageProcessor::readImage() {
    std::ifstream fileStream(inputFilename);

    if (!fileStream.is_open()) {
        throw std::runtime_error("Error opening file " + inputFilename);
    }

    image.reserve(NUM_ROWS_OR_NUM_PIXELS_PER_ROW);
    std::string line;
    int r_count = 0;

    while (std::getline(fileStream, line)) {
        image.emplace_back();
        auto& row = image.back();
        row.reserve(NUM_ROWS_OR_NUM_PIXELS_PER_ROW);

        std::stringstream ss(line);
        std::string pixel;
        int p_count = 0;

        while (std::getline(ss, pixel, ' ')) {
            std::stringstream ps(pixel);
            Color p{};
            char comma;

            if (!(ps >> p.r >> comma >> p.g >> comma >> p.b) || !p.isValid()) {
                throw std::runtime_error("Invalid pixel in line " + std::to_string(r_count));
            }

            row.push_back(p);
            p_count++;
        }
        if (p_count != NUM_ROWS_OR_NUM_PIXELS_PER_ROW) {
            throw std::runtime_error("Invalid amount of pixels in line " + std::to_string(r_count));
        }

        r_count++;
    }

    if (r_count != NUM_ROWS_OR_NUM_PIXELS_PER_ROW) {
        throw std::runtime_error("Invalid amount of lines in file");
    }

    fileStream.close();
}

void ImageProcessor::processImage(const Color& favoriteColor) {
    for (size_t row = 0; row < image.size(); ++row) {
        for (size_t col = 0; col < image[row].size(); ++col) {
            if (image[row][col] == favoriteColor) {
                if (row > 0) image[row - 1][col] = favoriteColor;
                if (col > 0) image[row][col - 1] = favoriteColor;
            }
        }
    }
}

void ImageProcessor::writeImage(const std::string& outputFilename) {
    std::ofstream file(outputFilename);

    for (const auto & row : image) {
        auto row_string = row | std::views::transform([](const Color& pixel) {
            return std::format("{},{},{}", pixel.r, pixel.g, pixel.b);
        }) | std::views::join_with(' ');

        file << std::ranges::to<std::string>(row_string) << std::endl;
    }
    file.close();
}
